### PR TITLE
Photino

### PR DESCRIPTION
**Pull request by victorporof**

### CODE REVIEW

**Comment by victorporof** [comment]
Switched to `socket.io` because `ws` couldn't be used in a browser environment, introduced a websocket connection between the frontend and browser server, and converted the small hello world frontend to use react.

Now actually working on transplanting the old frontend.

**Comment by ncalexan** [comment]
Does this implement Tofino issue 1414 (which I'm not hyperlinking to keep this slightly more private)?

**Comment by ncalexan** [comment]
As a general consideration, it would be easier to read and interpret this code if was split across fewer files.  I understand you're envisioning a scaffold on which to build many things, but it's easier to separate `index.js` into pieces later than it is to glue many small pieces together while reading.  The sense of the forest gets lost in the tiny trees.

**Comment by ncalexan** [comment]
I'm a little confused by the `browser-runner/` hierarchy.  That looks to be very dependent on Electron; would implementing a `qbrt`-based runner start by copying the `browser-runner/` hierarchy entirely, and then modifying things?  Or is there shared code across runners in there?

**Comment by ncalexan** [comment]
I've tried to digest a little of this work and I think I see the outlines.  I think it achieves a lot of what you set out to achieve.  Looking forward to seeing the next steps.

**Comment by victorporof** [comment]
> Does this implement Tofino issue 1414 (which I'm not hyperlinking to keep this slightly more private)?

In broad terms, yes, if the main point in 1414 is to communicate via standardized action format. The saga tangent is not strictly important, and app state syncing story isn't decided yet.

**Comment by victorporof** [comment]
> As a general consideration, it would be easier to read and interpret this code if was split across fewer files. I understand you're envisioning a scaffold on which to build many things, but it's easier to separate index.js into pieces later than it is to glue many small pieces together while reading. The sense of the forest gets lost in the tiny trees.

Agreed. I tried to reduce the number of files for now, but this will make a ton more sense once more of tofino gets transplanted.

**Comment by victorporof** [comment]
> I'm a little confused by the browser-runner/ hierarchy. That looks to be very dependent on Electron; would implementing a qbrt-based runner start by copying the browser-runner/ hierarchy entirely, and then modifying things? Or is there shared code across runners in there?

We discussed this briefly today. A better name would be electron-browser-runner, I expect us to have many such runners. There's no code sharing expected between implementations (although some minimal sharing could be possible).

**Comment by victorporof** [comment]
> I've tried to digest a little of this work and I think I see the outlines. I think it achieves a lot of what you set out to achieve. Looking forward to seeing the next steps.

@ncalexan let me know if there's anything that doesn't make sense. As discussed, the broad picture is encompassed in this PR, but we desperately need to figure out exact data flow.

File: src/browser-frontend/model/ui-page-model.js [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+  OPEN: 'open',
+  CLOSED: 'closed',
+};
+
+const UIPageModel = Immutable.Record({
+  tabState: TAB_STATES.INITIAL,
+  locationInputBarValue: '',
+});
+
+UIPageModel.TAB_STATES = TAB_STATES;
```
Comment (ncalexan) [inline src/browser-frontend/model/ui-page-model.js]: Is there an advantage in doing this in two steps?  Is it just that you want to refer to the value in the constructor, so you need some way to refer to it before the `Record` type is defined?

**Review by ncalexan — COMMENTED** [review]
I see that opening devtools runs through the browser server, so the flow would be:
- click in chrome UI
- chrome UI -> browser server (over WS)
- browser server -> browser runner (over WS)
- visible result in chrome UI

Is  there a reason to involve the BS here?  Is the idea that you want to be able to re-open devtools windows when you restart the BS?  This interaction feels like it could be within the chrome UI/BR process.  (Things like bookmark stars, of course, would flow across the high-latency browser runner -> browser server barrier.)

**Comment by victorporof** [comment]
> I see that opening devtools runs through the browser server [...]

I don't know the correct data flow in this case. My intuition tells me we want to avoid having all components being able to communicate with each other. Having a clearly defined data flow with the frontend on one end and the runner on the other, with the browser server as the main hub seems like a good idea, but this opinion is weakly held.

File: src/browser-frontend/model/ui-page-model.js (hunk 1 above)
Comment (victorporof) [reply src/browser-frontend/model/ui-page-model.js]: Yes to the latter.

**Comment by victorporof** [comment]
@fluffyemily @joewalker @rnewman @ncalexan This is ready for review if anyone is interested.

Do this first:
`git checkout bfg-browser` (this PR)
`git clean -f`
`npm install`

Run optimized build:
`npm start`

Run development build with logging enabled (opt-in because logging add lots of overhead in a web context with devtools open; for people hacking on UIs and CSS only, logging is irrelevant):
`npm run dev+logging`

Package:
`npm run package`

**Comment by victorporof** [comment]
At the very least if the browser-runner/browser-server innovation doesn't stick, we'll still end up with
1. Standardized communication over ws instead of ipc between the "platform" and the frontend
2. A dumb copy-pasteable frontend that is also easily hackable by the UX team
3. A photon implementation which we can potentially ship.
4. All of the knowledge we've gained trying this out :)

**Comment by victorporof** [comment]
As a final note: I expect merging the browser-runner and browser server into a thick electron main process to be relatively straightforward, should we end up wanting to do that.

**Comment by victorporof** [comment]
Next steps: quality of life improvements for the UX team, like hot module reload, rebuild on changes, using webpack dll plugin for faster build times.

**Comment by victorporof** [comment]
Marking this as closed, the bfg-browser branch is now the currently actively under development.

**Pull request closed without merging**